### ISSUE TITLE
Real-Time test coverage

### ISSUE THREAD

**Issue by gctucker** [kernelci/kernelci-core#474]
As mentioned at Plumbers 2020, `PREMPT_RT` is on its way to mainline:

https://linuxplumbersconf.org/event/7/contributions/766/

Having results for real-time testing in KernelCI would be very valuable and help with upstream support.  There is already a CI-RT system in place from Linutronix.  So one approach is to gather the results they already produce.  It's also possible to enable real-time tests to run on other existing CI systems such as CKI and gather their results too, and of course enable it as a native KernelCI support to run in the wide variety of hardware present in its existing distributed test labs.

At this stage, it's all about early discussion to determine what the next steps will be in practice as real-time support is progressively landing in mainline.

**Comment by igaw** [comment]
I'm doing stable-rt testing for the v4.4-rt branch with LAVA since a while. Doing so I learned a few things along. Let me add my lesson learned here.

Currently, I am using a (hackish) Python script for automating my workflow. The script is not really worth sharing, because it is not generic.

Anyway, the script does three main steps:
 1) Build several kernels with different configurations.  All configurations get only a boot test except the last configuration which gets a full rt-test suite run.

```
     CONFIG_NO_HZ=y
     CONFIG_SMP=n
     CONFIG_PREEMPT_NONE=y
     CONFIG_PREEMPT__LL=y
     CONFIG_PREEMPT_VOLUNTARY=y
     CONFIG_PREEMPT_RT_FULL=y (CONFIG_PREEMPT_RT=y)
``` 
    
 2) Create test jobs and trigger LAVA to run all jobs. The script generates a test job per rt-test program adding a background workload (currently just hackbench, planing to add some stress-ng workloads). For this I've extended LAVA's test-definitions, see  https://github.com/igaw/test-definitions/commits/preempt-rt
  Here is an example of one of the test jobs.
```
actions:
- deploy:
    kernel:
      type: bzImage
      url: http://lithium.lan:8080/artifacts/c2d/c2d-image-rt-v4.4.231-rt202-1-gd5f52c5bbf27
    nfsrootfs:
      url: http://lithium.lan:8080/artifacts/rootfs/debian-x86_64.tar
    os: debian
    timeout:
      minutes: 10
    to: tftp
- boot:
    auto_login:
      login_prompt: 'login:'
      username: root
    commands: nfs
    method: ipxe
    prompts:
    - 'root@debian:'
    timeout:
      minutes: 5
- test:
    definitions:
    - branch: preempt-rt
      from: git
      name: cyclictest
      parameters:
        AFFINTIY: 0-1
        BACKGROUND_CMD: hackbench
        DURATION: 300s
        INTERVAL: 1000
        MAX_LATENCY: 35
        PRIORITY: 98
        THREADS: 2
      path: automated/linux/cyclictest/cyclictest.yaml
      repository: https://github.com/igaw/test-definitions.git
    timeout:
      seconds: 360
device_type: x86
job_name: cyclictest
priority: medium
tags:
- c2d
timeouts:
  action:
    minutes: 30
  connection:
    minutes: 4
  job:
    minutes: 30
visibility: public
```

3) Fetch the results and dump it on my terminal. 

```
$ srt-build c2d jobs results 224                                                                                                                                                                                          224   0_cyclicdeadline     t1-max-latency      : fail    3042.00
  224   0_cyclicdeadline     t1-avg-latency      : pass    1708.00
  224   0_cyclicdeadline     t1-min-latency      : pass       1.00
  224   0_cyclicdeadline     t0-max-latency      : fail    3041.00
  224   0_cyclicdeadline     t0-avg-latency      : pass    1707.00
  224   0_cyclicdeadline     t0-min-latency      : pass       1.00
  225   0_cyclictest         t1-max-latency      : pass      12.00
  225   0_cyclictest         t1-avg-latency      : pass       2.00
  225   0_cyclictest         t1-min-latency      : pass       1.00
  225   0_cyclictest         t0-max-latency      : pass      12.00
  225   0_cyclictest         t0-avg-latency      : pass       2.00
  225   0_cyclictest         t0-min-latency      : pass       1.00
  226   0_pi-stress          pi-stress           : pass       0.00
  227   0_pmqtest            t3-2-max-latency    : pass      14.00
  227   0_pmqtest            t3-2-avg-latency    : pass       3.00
  227   0_pmqtest            t3-2-min-latency    : pass       2.00
  227   0_pmqtest            t1-0-max-latency    : pass      15.00
  227   0_pmqtest            t1-0-avg-latency    : pass       3.00
  227   0_pmqtest            t1-0-min-latency    : pass       2.00
  228   0_ptsematest         t3-2-max-latency    : pass      13.00
  228   0_ptsematest         t3-2-avg-latency    : pass       3.00
  228   0_ptsematest         t3-2-min-latency    : pass       2.00
  228   0_ptsematest         t1-0-max-latency    : pass      15.00
  228   0_ptsematest         t1-0-avg-latency    : pass       3.00
  228   0_ptsematest         t1-0-min-latency    : pass       2.00
  229   0_rt-migrate-test    t2-p53-avg          : pass      20.00
  229   0_rt-migrate-test    t2-p53-tot          : pass    1033.00
  229   0_rt-migrate-test    t2-p53-min          : pass       9.00
  229   0_rt-migrate-test    t2-p53-max          : pass      46.00
  229   0_rt-migrate-test    t1-p52-avg          : pass      23.00
  229   0_rt-migrate-test    t1-p52-tot          : pass    1151.00
  229   0_rt-migrate-test    t1-p52-min          : pass       8.00
  229   0_rt-migrate-test    t1-p52-max          : pass      55.00
  229   0_rt-migrate-test    t0-p51-avg          : pass   19224.00
  229   0_rt-migrate-test    t0-p51-tot          : pass  961206.00
  229   0_rt-migrate-test    t0-p51-min          : pass      13.00
  229   0_rt-migrate-test    t0-p51-max          : pass   20041.00
  230   0_signaltest         t0-max-latency      : pass      43.00
  230   0_signaltest         t0-avg-latency      : pass       8.00
  230   0_signaltest         t0-min-latency      : pass       4.00
  231   0_sigwaittest        t3-2-max-latency    : pass      16.00
  231   0_sigwaittest        t3-2-avg-latency    : pass       4.00
  231   0_sigwaittest        t3-2-min-latency    : pass       3.00
  231   0_sigwaittest        t1-0-max-latency    : pass      22.00
  231   0_sigwaittest        t1-0-avg-latency    : pass       4.00
  231   0_sigwaittest        t1-0-min-latency    : pass       3.00
  232   0_svsematest         t3-2-max-latency    : pass      14.00
  232   0_svsematest         t3-2-avg-latency    : pass       3.00
  232   0_svsematest         t3-2-min-latency    : pass       2.00
  232   0_svsematest         t1-0-max-latency    : pass      14.00
  232   0_svsematest         t1-0-avg-latency    : pass       3.00
  232   0_svsematest         t1-0-min-latency    : pass       2.00
  233   0_smoke-tests        linux-posix-lsb_release: pass
  233   0_smoke-tests        linux-posix-lscpu   : pass
  233   0_smoke-tests        linux-posix-ifconfig: pass
  233   0_smoke-tests        linux-posix-vmstat  : pass
  233   0_smoke-tests        linux-posix-pwd     : pass
  234   0_smoke-tests        linux-posix-lsb_release: pass
  234   0_smoke-tests        linux-posix-lscpu   : pass
  234   0_smoke-tests        linux-posix-ifconfig: pass
  234   0_smoke-tests        linux-posix-vmstat  : pass
  234   0_smoke-tests        linux-posix-uname   : pass
  234   0_smoke-tests        linux-posix-pwd     : pass
  235   0_smoke-tests        linux-posix-lsb_release: pass
  235   0_smoke-tests        linux-posix-lscpu   : pass
  235   0_smoke-tests        linux-posix-ifconfig: pass
  235   0_smoke-tests        linux-posix-vmstat  : pass
  235   0_smoke-tests        linux-posix-uname   : pass
  236   0_smoke-tests        linux-posix-lsb_release: pass
  236   0_smoke-tests        linux-posix-lscpu   : pass
  236   0_smoke-tests        linux-posix-ifconfig: pass
  236   0_smoke-tests        linux-posix-vmstat  : pass
  236   0_smoke-tests        linux-posix-uname   : pass
  236   0_smoke-tests        linux-posix-pwd     : pass
  237   0_smoke-tests        linux-posix-lsb_release: pass
  237   0_smoke-tests        linux-posix-lscpu   : pass
  237   0_smoke-tests        linux-posix-ifconfig: pass
  237   0_smoke-tests        linux-posix-vmstat  : pass
  237   0_smoke-tests        linux-posix-uname   : pass
  237   0_smoke-tests        linux-posix-pwd     : pass

```

Initially, I tried to teach LAVA to do all the result parsing (passed/failed) also for the performance numbers (see MAX_LATENCY) but it didn't work out as expected. So consider this a failed experiment. Properly the shell code in test-definition should be  removed again. 

So in short, I think most things are already there in LAVA, we just need to hook it up correctly. The background command feature is a bit simple though. As Mel Gorman pointed out, we might want to have more sophisticated ways to do performance measurements (see https://github.com/gormanm/mmtests). Nevertheless, this simple background command feature could just be a starting point. Getting LAVA to run mmtests might needs a bit more thinking first.

**Comment by khilman** [comment]
@igaw Can you give a bit more details on how you build your debian rootfs?  We should be able to pretty easily convert that into a KernelCI rootfs config (c.f. https://github.com/kernelci/kernelci-core/blob/master/rootfs-configs.yaml for examples of how rootfs for other tests are configured)

**Comment by moto-timo** [comment]
FWIW, we wrapped cyclictest with pytest (LAVA would just run pytest) and launched hackbench with a systemd service, but those are irrelevant implementation details.

**Comment by khilman** [comment]
> FWIW, we wrapped cyclictest with pytest (LAVA would just run pytest) and launched hackbench with a systemd service, but those are irrelevant implementation details.

@moto-timo your LAVA changes to use pytest would be very interesting to have in upstream LAVA.  Could greatly expand the types of tests we could run.

**Comment by khilman** [comment]
@igaw Using our debian rootfs and your test-defintions, I was able to create the build-configs and test plans to run your setup in KernelCI.  (c.f. https://github.com/kernelci/kernelci-core/pull/475)

One thing I notice though is that cyclictest output is made for interactive terimanls, but when run on a dumb terminal makes for huge LAVA logs.   

I'm currently just using cyclictest from the debian rt-tests package.  Is there a way to make your test-definition make the stdout less noisy/cluttered?

**Comment by khilman** [comment]
FYI: first batch of results from manually run tests on the staging backend: https://staging.kernelci.org/test/plan/id/5f4581c91902afaf5f544a7b/

**Comment by moto-timo** [comment]
@khilman actually, what we did was a pytest reporting plugin to spit out what LAVA wanted to see. Let me revisit getting that through the internal Open Source process and publish it on pypi.

**Comment by igaw** [comment]
@khilman nice! Don't worry about the hardware selection at this point. I suppose those things can be sorted out after the test suite is setup correctly. I am testing v4.4-rt only on an old x86_64 box and a BeagleBone Black. The ARMv8 support is there in v4.4.

**rootfs**

I came up with two scripts for creating the Debian rootfs. The recommended Debian way is to use the d-i installer. This was my
first version. But it was so slow I got tired of it and tried debootstrap. Although not supported it works pretty well for me.

https://www.monom.org/data/preempt_rt/create-rootfs-debian-installer.sh
https://www.monom.org/data/preempt_rt/create-rootfs-debootstrap.sh

**cyclictest output**

Indeed, the verbosity is a bit annoying. The test-definitions testsuite explicitly says no test should be made quiet, https://docs.lavasoftware.org/lava/writing-tests.html#test-writer-scripts Though I think the rt-tests are bit special in this regard. I'll try to get the silence knobs added upstream. Also enabling the histogram makes sense.

**Comment by igaw** [comment]
My Python script does a 'make defconfig' first and then uses the merge_config.sh script to add the missing config snippets. 

I've uploaded my script here: https://www.monom.org/data/preempt_rt/srt-build
Don't know if this helps as it is ugly as hell and very specific for my setup. The interesting part is the 'lava' command :)

For example for the flavor 'rt', it's 'make default' + config_base + config-base-rt + config-rcu + rt/config-rt-trace

The config snippets: https://www.monom.org/data/preempt_rt/configs/
and the job templates: https://www.monom.org/data/preempt_rt/jobs/

**Comment by gctucker** [comment]
The first results are in: https://staging.kernelci.org/job/rt-stable_v5.4-rt/branch/HEAD/

Thanks to @khilman for the plumbing :) #475

**Comment by khilman** [comment]
@igaw thanks for all the details.

Re: cyclictest verbosity, I submiited a quick fix to upstream already for the quiet part: https://github.com/Linaro/test-definitions/pull/199
But I'll let you work out the histogram part since I'm not sure about the sane defautls for the -h parameter.

So whenever your `preempt-rt` branch pulls in that change, we'll have quieter output.  The current test plan I submitted is using your fork of test-definitions, not upstream.  Let me know if you prefer that we use upstream.

**Comment by igaw** [comment]
@khilman I see that your quick fix got merged. I'll rebase my branch and also add the '-q' options to the other rt-tests, e.g. cyclicdeadline. Then I'll send out a pull request. I think we should have all in upstream eventually. 

The reason why I kept my changes locally for a while was to test them and get a feeling if this is the right thing to do. The MAX_LATENCY feature was not really though through.

**Comment by igaw** [comment]
@khilman can I somehow help to get more test hooked up and eventually more stable-rt trees added?

**Comment by gctucker** [comment]
@igaw To add a new branch:
https://github.com/kernelci/kernelci-core/issues/new?assignees=&labels=&template=new-kernel-branch.md&title=Add+branch+BRANCH+from+TREE

**Comment by gctucker** [comment]
Either that or create a PR to add it in `build-configs.yaml`.
